privilege/privileges/privileges.go

```diff
@@ -0,0 +1,266 @@
+	if ok {
+		return true, nil
+	}
+	// Check db scope privileges.
+	dbp, ok := p.privs.DBPrivs[db.Name.O]
+	if ok {
+		ok = dbp.contain(privilege)
+		if ok {
+			return true, nil
+		}
+	}
+	if tbl == nil {
+		return false, nil
+	}
+	// Check table scope privileges.
+	dbTbl, ok := p.privs.TablePrivs[db.Name.O]
+	if !ok {
+		return false, nil
+	}
+	tblp, ok := dbTbl[tbl.Name.O]
+	if !ok {
+		return false, nil
+	}
+	return tblp.contain(privilege), nil
+}
+
+func (p *UserPrivileges) loadPrivileges(ctx context.Context) error {
+	user := variable.GetSessionVars(ctx).User
+	strs := strings.Split(user, "@")
+	p.username, p.host = strs[0], strs[1]
+	p.privs = &userPrivileges{}
+	// Load privileges from mysql.User/DB/Table_privs/Column_privs table
+	err := p.loadGlobalPrivileges(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	err = p.loadDBScopePrivileges(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	err = p.loadTableScopePrivileges(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	// TODO: consider column scope privilege latter.
+	return nil
+}
+
+// mysql.User/mysql.DB table privilege columns start from index 3.
+// See: booststrap.go CreateUserTable/CreateDBPrivTable
+const userTablePrivColumnStartIndex = 3
+const dbTablePrivColumnStartIndex = 3
+
+func (p *UserPrivileges) loadGlobalPrivileges(ctx context.Context) error {
+	sql := fmt.Sprintf(`SELECT * FROM %s.%s WHERE User="%s" AND (Host="%s" OR Host="%%");`, mysql.SystemDB, mysql.UserTable, p.username, p.host)
+	rs, err := ctx.(sqlexec.RestrictedSQLExecutor).ExecRestrictedSQL(ctx, sql)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer rs.Close()
+	ps := &privileges{}
+	fs, err := rs.Fields()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	for {
+		row, err := rs.Next()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if row == nil {
+			break
+		}
+		for i := userTablePrivColumnStartIndex; i < len(fs); i++ {
+			d := row.Data[i]
+			ed, ok := d.(mysql.Enum)
+			if !ok {
+				return errors.Errorf("Privilege should be mysql.Enum: %v(%T)", d, d)
+			}
+			if ed.String() != "Y" {
+				continue
+			}
+			f := fs[i]
+			p, ok := mysql.Col2PrivType[f.Name]
+			if !ok {
+				panic("Unknown Privilege Type!")
+			}
+			ps.add(p)
+		}
+	}
+	p.privs.GlobalPrivs = ps
+	return nil
+}
+
+func (p *UserPrivileges) loadDBScopePrivileges(ctx context.Context) error {
+	sql := fmt.Sprintf(`SELECT * FROM %s.%s WHERE User="%s" AND (Host="%s" OR Host="%%");`, mysql.SystemDB, mysql.DBTable, p.username, p.host)
+	rs, err := ctx.(sqlexec.RestrictedSQLExecutor).ExecRestrictedSQL(ctx, sql)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer rs.Close()
+	ps := make(map[string]*privileges)
+	fs, err := rs.Fields()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	for {
+		row, err := rs.Next()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if row == nil {
+			break
+		}
+		// DB
+		db, ok := row.Data[1].(string)
+		if !ok {
+			panic("This should be never happened!")
```

Comment:
error means expected excption. But if ok is false, there must be something wrong. And it is a fatal error.